source/common/router/header_formatter.cc

```diff
@@ -2,13 +2,109 @@
+#include "common/config/metadata.h"
+#include "common/json/json_loader.h"
 
 #include "fmt/format.h"
 
 namespace Envoy {
 namespace Router {
 
+namespace {
+
+std::string formatUpstreamMetadataParseException(const std::string& params,
+                                                 const EnvoyException* cause = nullptr) {
+  std::string reason;
+  if (cause != nullptr) {
+    reason = fmt::format(", because {}", cause->what());
+  }
+
+  return fmt::format("Incorrect header configuration. Expected format "
+                     "UPSTREAM_METADATA([\"namespace\", \"k\", ...]), actual format "
+                     "UPSTREAM_METADATA{}{}",
+                     params, reason);
+}
+
+std::function<std::string(const Envoy::AccessLog::RequestInfo&)>
+parseUpstreamMetadataField(const std::string& params_str) {
+  if (params_str.empty() || params_str[0] != '(' || params_str[params_str.size() - 1] != ')') {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str));
+  }
+
+  std::vector<std::string> params;
+  try {
+    Json::ObjectSharedPtr parsed_params =
+        Json::Factory::loadFromString(StringUtil::subspan(params_str, 1, params_str.size() - 1));
+
+    for (const auto& param : parsed_params->asObjectArray()) {
+      params.emplace_back(param->asString());
+    }
+  } catch (Json::Exception& e) {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str, &e));
+  }
+
+  // Minimum parameters are a metadata namespace (e.g. "envoy.lb") and a metadata key.
+  if (params.size() < 2) {
+    throw EnvoyException(formatUpstreamMetadataParseException(params_str));
+  }
+
+  return [params](const Envoy::AccessLog::RequestInfo& request_info) -> std::string {
+    Upstream::HostDescriptionConstSharedPtr host = request_info.upstreamHost();
+    if (!host) {
+      return "";
+    }
+
+    const ProtobufWkt::Value* value =
+        &Config::Metadata::metadataValue(host->metadata(), params[0], params[1]);
+    if (value->kind_case() == ProtobufWkt::Value::KIND_NOT_SET) {
+      // No kind indicates default ProtobufWkt::Value which means namespace or key not
+      // found.
+      return "";
+    }
+
+    size_t i = 2;
+    while (i < params.size()) {
+      if (!value->has_struct_value()) {
+        break;
+      }
+
+      const auto field_it = value->struct_value().fields().find(params[i]);
+      if (field_it == value->struct_value().fields().end()) {
+        return "";
+      }
+
+      value = &field_it->second;
+      i++;
+    }
+
+    if (i < params.size()) {
+      // Didn't find all the keys.
+      return "";
+    }
+
+    switch (value->kind_case()) {
+    case ProtobufWkt::Value::kNumberValue:
+      return fmt::format("{}", value->number_value());
+
+    case ProtobufWkt::Value::kStringValue:
+      return value->string_value();
+
+    case ProtobufWkt::Value::kBoolValue:
+      return value->bool_value() ? "true" : "false";
+
+    default:
+      // Unsupported type or null value.
```

Comment:
This case will be hard for a user to debug.  Since we found something, but it's the wrong type, would it be better to return "unsupported metadata type" or something, or perhaps log something?

Reply:
I think returning an arbitrary string would be bad. I've added a log message.

Reply:
IMO debug is a more appropriate log level here, as this could emit at high rate due to external factors.